### PR TITLE
feat(webcal): only update modified and deleted events from webcal calendars

### LINKED ISSUE #43541: Use Delta for subscription calendars instead of deleting and reinserting all events
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

https://github.com/nextcloud/server/pull/42345 / https://github.com/nextcloud/server/pull/43261 shows a long running transaction for the webcal background sync:

```
Exception: Long running transaction
#9 /var/www/nextcloud/lib/private/DB/Connection.php(680): OC\DB\Connection::commit
#8 /var/www/nextcloud/lib/private/DB/ConnectionAdapter.php(154): OC\DB\ConnectionAdapter::commit
#7 /var/www/nextcloud/lib/public/AppFramework/Db/TTransactional.php(64): OCA\DAV\CalDAV\CalDavBackend::atomic
#6 /dav/lib/CalDAV/CalDavBackend.php(3075): OCA\DAV\CalDAV\CalDavBackend::purgeAllCachedEventsForSubscription
#5 /dav/lib/CalDAV/WebcalCaching/RefreshWebcalService.php(101): OCA\DAV\CalDAV\WebcalCaching\RefreshWebcalService::refreshSubscription
#4 /dav/lib/BackgroundJob/RefreshWebcalJob.php(115): OCA\DAV\BackgroundJob\RefreshWebcalJob::run
#3 /var/www/nextcloud/lib/public/BackgroundJob/Job.php(81): OCP\BackgroundJob\Job::start
#2 /var/www/nextcloud/lib/public/BackgroundJob/Job.php(67): OCP\BackgroundJob\Job::execute
#1 /dav/lib/BackgroundJob/RefreshWebcalJob.php(108): OCA\DAV\BackgroundJob\RefreshWebcalJob::execute
#0 /var/www/nextcloud/cron.php(152): null
```

### Steps to reproduce

1. Set up Nextcloud 28.0.2
2. Install Calendar
3. Add a webcal subscription

### Expected behavior

Fast transaction

### Installation method

None

### Nextcloud Server version

28

### Operating system

None

### PHP engine version

None

### Web server

None

### Database engine version

None

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

_No response_

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by miaulalala**
* Resolves: https://github.com/nextcloud/server/issues/43541

## Summary
Blackfire profiling comparison for an update query:
image

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] ~~Screenshots before/after for front-end changes~~
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required - https://github.com/nextcloud/documentation/pull/12097
- [ ] ~~[Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)~~

### CODE REVIEW

File: apps/dav/lib/CalDAV/WebcalCaching/RefreshWebcalService.php [hunk 1]
```diff
@@ -304,47 +232,24 @@ private function updateSubscription(array $subscription, array $mutations) {
...
-	}
-
-	/**
-	 * Returns a random uri for a calendar-object
-	 *
-	 * @return string
-	 */
-	public function getRandomCalendarObjectUri():string {
-		return UUIDUtil::getUUID() . '.ics';
+		return md5($localVobject->serialize()) === md5($vObject->serialize());
```
Comment (ChristophWurst) [inline apps/dav/lib/CalDAV/WebcalCaching/RefreshWebcalService.php:253]: Do we actually need to hash the strings or can we compare directly?

Comment (miaulalala) [reply apps/dav/lib/CalDAV/WebcalCaching/RefreshWebcalService.php:253]: good point, I think strcasecmp should work too

**Comment by ChristophWurst** [comment]
What does the blackfire comparison look like without md5? Does it get even better?

**Review by ChristophWurst — APPROVED** [review]
Looks good!

**Comment by miaulalala** [comment]
> What does the blackfire comparison look like without md5? Does it get even better?
md5 vs strcasecmp:
image

**Review by ChristophWurst — APPROVED** [review]
Code looks good

:shipit:

**Review by st3iny — APPROVED** [review]
Tested and works.

**Comment by st3iny** [comment]
I ran the freeze check again and we are good.

**Pull request merged**